src/multidoc-manager.js

```diff
@@ -0,0 +1,567 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CommonSignals} from './common-signals';
+import {Services} from './services';
+import {VisibilityState} from './visibility-state';
+import {childElementsByTag, isConnectedNode} from './dom';
+import {
+  createShadowDomWriter,
+  createShadowRoot,
+  importShadowBody,
+} from './shadow-embed';
+import {dev, user} from './log';
+import {disposeServicesForDoc, getServicePromiseOrNullForDoc} from './service';
+import {getMode} from './mode';
+import {installStylesForDoc} from './style-installer';
+import {isArray, isObject} from './types';
+import {parseUrlDeprecated} from './url';
+import {setStyle} from './style';
+
+/** @const @private {string} */
+const TAG = 'multidoc-manager';
+
+/**
+ * A manager for documents in the multi-doc environment.
+ */
+export class MultidocManager {
+  /**
+   * @param {!Window} win
+   * @param {!./service/ampdoc-impl.AmpDocService} ampdocService
+   * @param {!./service/extensions-impl.Extensions} extensions
+   * @param {!./service/timer-impl.Timer} timer
+   */
+  constructor(win, ampdocService, extensions, timer) {
+    /** @const */
+    this.win = win;
+    /** @private @const */
+    this.ampdocService_ = ampdocService;
+    /** @private @const */
+    this.extensions_ = extensions;
+    /** @private @const */
+    this.timer_ = timer;
+
+    /** @private @const {!Array<!ShadowRoot>} */
+    this.shadowRoots_ = [];
+  }
+
+  /**
+   * Attaches the shadow root and calls the supplied DOM builder.
+   * @param {!Element} hostElement
+   * @param {string} url
+   * @param {!Object<string, string>|undefined} params
+   * @param {function(!Object, !ShadowRoot,
+   * !./service/ampdoc-impl.AmpDocShadow):!Promise} builder
+   * @return {!ShadowDoc}
+   * @private
+   */
+  attachShadowDoc_(hostElement, url, params, builder) {
+    params = params || Object.create(null);
+    this.purgeShadowRoots_();
+
+    setStyle(hostElement, 'visibility', 'hidden');
+    const shadowRoot = createShadowRoot(hostElement);
+
+    // TODO: closeShadowRoot_ is asynchronous. While this safety check is well
+    // intentioned, it leads to a race between unlayout and layout of custom
+    // elements.
+    if (shadowRoot.AMP) {
+      user().warn(TAG, "Shadow doc wasn't previously closed");
+      this.closeShadowRoot_(shadowRoot);
+    }
+
+    const amp = {};
+    shadowRoot.AMP = amp;
+    amp.url = url;
+    const {origin} = parseUrlDeprecated(url);
+
+    const ampdoc = this.ampdocService_.installShadowDoc(url, shadowRoot, {
+      params,
+    });
+    /** @const {!./service/ampdoc-impl.AmpDocShadow} */
+    amp.ampdoc = ampdoc;
+    dev().fine(TAG, 'Attach to shadow root:', shadowRoot, ampdoc);
+
+    // Install runtime CSS.
+    installStylesForDoc(
```

Comment:
Doing a diff, these two lines seem to be the only changes? I'm astounded that this was all that's needed.

Reply:
Yup! I essentially did a diff of `multidoc-manager` dependency tree and `runtime`'s dependency tree to see what `multidoc-manager` needs that's not provided by `runtime`. I'll do more testing to make sure this doesn't break anything, otherwise I'll just exclude the `inabox` runtime from this change and we should be good to go.